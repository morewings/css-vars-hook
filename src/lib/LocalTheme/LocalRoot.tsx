import type {HTMLAttributes, ReactNode} from 'react';
import {createElement, forwardRef, useEffect, useRef} from 'react';

import {createStyleObject} from '../utils';

export type LocalRootProps<TTheme extends Record<string, string> = Record<string, string>> = JSX.IntrinsicAttributes &
    HTMLAttributes<HTMLElement> & {
        children?: ReactNode;
        /** Apply your own CSS class or use styled component styled(LocalRoot) */
        className?: string;
        /** Choose which HTMLElement to render as a root. div is default. */
        as?: string;
        /** Apply initial theme. */
        theme?: TTheme;
        /** Provide theme setter function. */
        setTheme?: (arg0: TTheme) => void;
    };

export const LocalRoot = forwardRef<HTMLElement, LocalRootProps>((props, ref) => {
    // This is needed to fix an error introduced in version 0.6.14.
    // Props were not transported to returned HTMLElement.
    const {children, as = 'div', theme, setTheme = () => {}, ...restProps} = props;

    const initialStyle = useRef(createStyleObject(theme));

    useEffect(() => {
        setTheme(theme);
    }, [theme, setTheme]);

    return createElement(as, {...restProps, style: initialStyle.current, ref}, children);
});
