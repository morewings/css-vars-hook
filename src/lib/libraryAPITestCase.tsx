import type {FC} from 'react';
import {createRef} from 'react';
import {render} from '@testing-library/react';

/**
 * Test utility containing a set of tests to cover Library API features.
 * @see https://storybook.olxtech.pl/?path=/docs/experimental-library-api--page
 * @see LibraryProps
 * @example
 * // inside test file
 * libraryAPITestCase(
 *  // provide a Component to test
 *  Accordion,
 *  // Set name to label tests
 *  'Accordion',
 *  // Add props required to run component
 *  { content: itemSingleComponent }
 *  );
 */
export const libraryAPITestCase = <TProps,>(
    Component: FC<TProps>,
    name: string,
    defaultProps = {} as TProps
) => {
    describe(`${name} - Native API`, () => {
        const propValue = 'bar';

        it('renders provided `data-` attributes', () => {
            const name = 'foo';

            const {getByTestId} = render(
                <Component
                    {...defaultProps}
                    data-foo={propValue}
                    data-testid={propValue}
                    className={propValue}
                />
            );
            expect(getByTestId(propValue)).toHaveAttribute(`data-${name}`, propValue);
            expect(getByTestId(propValue)).toHaveAttribute('data-testid', propValue);
        });

        it.each(['id', 'role'])('renders native string props', propName => {
            const {getByTestId} = render(
                <Component
                    {...defaultProps}
                    data-testid={propValue}
                    {...{[propName]: propValue}}
                />
            );
            expect(getByTestId(propValue)).toHaveAttribute(propName, propValue);
        });

        it('renders `className` prop', () => {
            const {getByTestId} = render(
                <Component
                    {...defaultProps}
                    data-testid={propValue}
                    className={propValue}
                />
            );
            expect(getByTestId(propValue).getAttribute('class')).toMatch(propValue);
        });

        it('accepts ref as a callback', () => {
            const ref = jest.fn();
            render(<Component {...defaultProps} ref={ref} />);
            expect(ref.mock.calls[0][0]).toMatchSnapshot();
        });

        it('accepts ref as a Mutable Object', () => {
            const ref = createRef<HTMLInputElement>();
            render(<Component {...defaultProps} ref={ref} />);
            expect(ref.current).toMatchSnapshot();
        });
    });
};
