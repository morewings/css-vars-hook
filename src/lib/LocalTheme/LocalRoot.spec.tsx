import {render} from '@testing-library/react';

import {libraryAPITestCase} from '@/lib/libraryAPITestCase.tsx';

import {LocalRoot} from './LocalRoot.tsx';

const themeFoo = {
    foo: 'foo',
};
const themeBar = {
    bar: 'bar',
};

describe('lib > LocalRoot', () => {
    it('renders children', () => {
        const children = 'foo';
        const {getByText} = render(<LocalRoot>{children}</LocalRoot>);
        expect(getByText(children)).toBeInTheDocument();
    });

    it('respects `as` prop', () => {
        const {rerender, asFragment} = render(<LocalRoot as="button" />);
        expect(asFragment()).toMatchSnapshot();
        rerender(<LocalRoot as="main" />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('creates initial style based on `theme` prop', () => {
        const {getByTestId} = render(
            <LocalRoot data-testid="provider" theme={themeFoo} />
        );
        expect(getByTestId('provider').getAttribute('style')).toMatchSnapshot();
    });

    it('sets and updates initial style, using `setTheme` callback when `theme` prop changes', () => {
        const setTheme = jest.fn();
        const {rerender, getByTestId} = render(
            <LocalRoot data-testid="provider" theme={themeFoo} setTheme={setTheme} />
        );
        expect(setTheme).toHaveBeenCalledTimes(1);
        expect(setTheme).toHaveBeenCalledWith(themeFoo);
        expect(getByTestId('provider').getAttribute('style')).toMatchSnapshot();
        rerender(
            <LocalRoot data-testid="provider" theme={themeBar} setTheme={setTheme} />
        );
        expect(setTheme).toHaveBeenCalledTimes(2);
        expect(setTheme).toHaveBeenCalledWith(themeBar);
        expect(getByTestId('provider').getAttribute('style')).toMatchSnapshot();
    });

    libraryAPITestCase(LocalRoot, 'LocalRoot');
});
