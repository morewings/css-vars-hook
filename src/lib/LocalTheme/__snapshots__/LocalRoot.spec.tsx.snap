// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib > LocalRoot LocalRoot - Native API accepts ref as a Mutable Object 1`] = `<div />`;

exports[`lib > LocalRoot LocalRoot - Native API accepts ref as a callback 1`] = `<div />`;

exports[`lib > LocalRoot creates initial style based on \`theme\` prop 1`] = `"--foo: foo;"`;

exports[`lib > LocalRoot respects \`as\` prop 1`] = `
<DocumentFragment>
  <button />
</DocumentFragment>
`;

exports[`lib > LocalRoot respects \`as\` prop 2`] = `
<DocumentFragment>
  <main />
</DocumentFragment>
`;

exports[`lib > LocalRoot sets and updates initial style, using \`setTheme\` callback when \`theme\` prop changes 1`] = `"--foo: foo;"`;

exports[`lib > LocalRoot sets and updates initial style, using \`setTheme\` callback when \`theme\` prop changes 2`] = `"--bar: bar;"`;
